cmd/kaspaminer/mineloop.go

```diff
@@ -132,48 +131,67 @@ func handleFoundBlock(client *minerClient, block *externalapi.DomainBlock) error
 	return nil
 }
 
-func solveBlock(block *externalapi.DomainBlock, stopChan chan struct{}, foundBlock chan *externalapi.DomainBlock) {
-	targetDifficulty := difficulty.CompactToBig(block.Header.Bits())
-	headerForMining := block.Header.ToMutable()
+func mineNextBlock(mineWhenNotSynced bool) *externalapi.DomainBlock {
 	initialNonce := rand.Uint64() // Use the global concurrent-safe random source.
-	for i := initialNonce; i != initialNonce-1; i++ {
-		select {
-		case <-stopChan:
-			return
-		default:
-			headerForMining.SetNonce(i)
-			atomic.AddUint64(&hashesTried, 1)
-			if pow.CheckProofOfWorkWithTarget(headerForMining, targetDifficulty) {
-				block.Header = headerForMining.ToImmutable()
-				foundBlock <- block
-				return
-			}
+	for i := initialNonce; ; i++ {
+		block := getBlockForMining(mineWhenNotSynced)
+		targetDifficulty := difficulty.CompactToBig(block.Header.Bits())
+		headerForMining := block.Header.ToMutable()
+		headerForMining.SetNonce(i)
+		atomic.AddUint64(&hashesTried, 1)
+		if pow.CheckProofOfWorkWithTarget(headerForMining, targetDifficulty) {
+			block.Header = headerForMining.ToImmutable()
+			log.Infof("Found block %s with parents %s", consensushashing.BlockHash(block), block.Header.ParentHashes())
+			return block
 		}
 	}
 }
 
-func templatesLoop(client *minerClient, miningAddr util.Address,
-	newTemplateChan chan *appmessage.GetBlockTemplateResponseMessage, errChan chan error, stopChan chan struct{}) {
+func getBlockForMining(mineWhenNotSynced bool) *externalapi.DomainBlock {
```

Comment:
@someone235 I must say that all this complication leads me to believe that a single channel with a for+switch on it would be simpler